File: packages/php/tests/PostgresStorageTest.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use Greyfoundry\Identifold\Storage\Postgres\PostgresStorageAdapter;
use Greyfoundry\Identifold\Storage\ReferenceReservation;
use Greyfoundry\Identifold\Storage\SequenceAllocationRequest;

$databaseUrl = getenv('IDENTIFOLD_TEST_DATABASE_URL');
if ($databaseUrl === false) {
    exit(0);
}
$parts = parse_url($databaseUrl);
if (!is_array($parts)) {
    throw new RuntimeException('database URL');
}
$pdo = new PDO(
    sprintf(
        'pgsql:host=%s;port=%d;dbname=%s',
        $parts['host'],
        $parts['port'],
        ltrim($parts['path'], '/'),
    ),
    $parts['user'],
    $parts['pass'],
    [PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION],
);
$root = dirname(__DIR__, 3);
foreach ([
    '001_identifold.down.sql',
    '001_identifold.up.sql',
    '003_idempotent_replay.up.sql',
    '004_reference_lookup.up.sql',
] as $migration) {
    $pdo->exec(file_get_contents(
        $root . '/integrations/postgres/migrations/' . $migration,
    ));
}

$adapter = new PostgresStorageAdapter($pdo);
$randomMid = '01890f8c-7b2a-7cc3-98b0-112233445566';
$randomRef = 'ORD-0123-4567-89-P';
assert($adapter->reserve(new ReferenceReservation($randomMid, 'order', $randomRef)));
assert($adapter->resolve($randomRef, 'order')?->machineId === $randomMid);
$request = new SequenceAllocationRequest(
    '01890f8c-7b2a-7cc3-98b0-112233445567',
    'receipt',
    'RCT',
    null,
    4,
);
assert($adapter->allocate($request) === 1);
assert($adapter->allocate($request) === 1);
assert($adapter->resolve('RCT-0001-1', 'receipt')?->machineId === $request->machineId);

File: packages/ruby/test/postgres_storage_test.rb
require "minitest/autorun"
require "pg"
require_relative "../lib/identifold"
require_relative "../lib/identifold/storage/postgres"

class PostgresStorageTest < Minitest::Test
  def test_reserves_allocates_and_resolves
    database_url = ENV["IDENTIFOLD_TEST_DATABASE_URL"]
    skip "IDENTIFOLD_TEST_DATABASE_URL is not configured" unless database_url

    connection = PG.connect(database_url)
    root = File.expand_path("../../..", __dir__)
    %w[
      001_identifold.down.sql
      001_identifold.up.sql
      003_idempotent_replay.up.sql
      004_reference_lookup.up.sql
    ].each do |migration|
      connection.exec(File.read(File.join(root, "integrations/postgres/migrations", migration)))
    end

    adapter = Identifold::Storage::PostgresAdapter.new(connection)
    random_mid = "01890f8c-7b2a-7cc3-98b0-112233445566"
    random_ref = "ORD-0123-4567-89-P"
    assert adapter.reserve(Identifold::Storage::ReferenceReservation.new(
      machine_id: random_mid, namespace: "order", reference: random_ref
    ))
    assert_equal random_mid, adapter.resolve(random_ref, "order").machine_id

    request = Identifold::Storage::SequenceAllocationRequest.new(
      machine_id: "01890f8c-7b2a-7cc3-98b0-112233445567",
      namespace: "receipt",
      reference_prefix: "RCT",
      scope: nil,
      width: 4
    )
    assert_equal 1, adapter.allocate(request)
    assert_equal 1, adapter.allocate(request)
    assert_equal request.machine_id, adapter.resolve("RCT-0001-1", "receipt").machine_id
  ensure
    connection&.close
  end
end
